kvserver: add datadriven test for kvflowcontrol integration

irfansharif opened the pull request:
(First six commits are from #98308.)

Part of #95563. We add TestFlowControlIntegration, exercising the kvflowcontrol integration interface (replicaFlowControlIntegration) introduced in an earlier commit. It tests how the underlying kvflowcontrol.Handle is constructed/destroyed, and its stream-management APIs invoked, as replicas acquire raft leadership, lost it, observe paused and/or inactive followers, change range descriptors, and react to raft progress updates. We can write tests of the following form:

    # Observe how the integration layer deals with paused followers.
    # Start off with a triply replicated range r1/t1, with replicas on
    # n1/s1, n2/s2, and n3/s3 (with replica IDs 1-3 respectively).
    init tenant=t1 range=r1 replid=1
    ----

    state descriptor=(1,2,3) applied=1/10
    ----

    # Set up replid=1 (declared in init above) to be the raft leader. It
    # should connect to all three replication streams.
    integration op=became-leader
    ----
    initialized flow control handle for r1/t1
    connected to replication stream t1/s1 starting at log-position=1/10
    connected to replication stream t1/s2 starting at log-position=1/10
    connected to replication stream t1/s3 starting at log-position=1/10

    # Pause replid=2. Observe that we disconnect the stream to t1/s2.
    state descriptor=(1,2,3) paused=(2)
    ----

    integration op=followers-paused
    ----
    disconnected from replication stream t1/s2

These are still unit tests, testing the interface at just the interface-level. We'll introduce more end-to-end integration testing of an actual replica's lifecycle in subsequent commits.

Release note: None

irfansharif committed:
rpc: remove debug logging

This was added recently, in #94778, and contributes to log spam of the
following sort:

I230404 15:00:33.826337 2400 rpc/context.go:2249  [T1,n1,rnode=2,raddr=127.0.0.1:55941,class=default,rpc] 268  connection heartbeat loop ended with err: <nil>
I230404 15:00:33.826338 3986 rpc/context.go:2249  [T1,n2,rnode=3,raddr=127.0.0.1:55955,class=system,rpc] 269  connection heartbeat loop ended with err: <nil>
I230404 15:00:33.826367 3455 rpc/context.go:2249  [T1,n2,rnode=3,raddr=127.0.0.1:55955,class=default,rpc] 270  connection heartbeat loop ended with err: <nil>
I230404 15:00:33.826394 3354 rpc/context.go:2249  [T1,n2,rnode=2,raddr=127.0.0.1:55941,class=default,rpc] 271  connection heartbeat loop ended with err: <nil>

Release note: None

irfansharif committed:
inspectz,kvflowinspectpb: add /inspectz pages

This commit adds rudimentary /inspectz-style pages to CRDB. These hang
off of <url>/inspectz and support a few registered endpoints to inspect
kvflowcontrol state in JSON form.

  - /inspectz/kvflowcontroller: marshals the state of registered flow
    control streams, and how many {regular,elastic} tokens are available
    for each one.

  - /inspectz/kvflowhandles: marshals the state of all in-memory
    kvflowcontrol.Handles held per leader+leaseholder replica, showing
    all deducted tokens and corresponding log positions, and what
    stream(s) each replica is connected to. It also supports querying
    for specific ranges through a ?ranges=<int>[,<int>] query parameter.

To power these endpoints we introduced proto representations for various
components under pkg/../kvflowcontrol/kvflowinspectpb. Select
kvflowcontrol interfaces now expose an Inspect() method, returning the
relevant proto.

Other than just observing proto-state like the 90s, this commit also
adds indexed crdb_internal vtables to do more sophisticated filtering of
these inspectz-protos. It's easy to combine these tables to understand
exactly which ranges are blocked on flow tokens, and for which streams
in particular. They are:

    CREATE TABLE crdb_internal.kv_flow_controller (
      tenant_id                INT NOT NULL,
      store_id                 INT NOT NULL,
      available_regular_tokens INT NOT NULL,
      available_elastic_tokens INT NOT NULL
    )

    CREATE TABLE crdb_internal.kv_flow_control_handles (
      range_id                 INT NOT NULL,
      tenant_id                INT NOT NULL,
      store_id                 INT NOT NULL,
      total_tracked_tokens     INT NOT NULL,
      INDEX(range_id)
    )

    CREATE TABLE crdb_internal.kv_flow_token_deductions (
      range_id  INT NOT NULL,
      tenant_id INT NOT NULL,
      store_id  INT NOT NULL,
      priority  STRING NOT NULL,
      log_term  INT NOT NULL,
      log_index INT NOT NULL,
      tokens    INT NOT NULL,
      INDEX(range_id)
    )

To see the set of ranges blocked on regular flow tokens, one can run
something like:

    SELECT range_id,
           crdb_internal.humanize_bytes(available_regular_tokens)
      FROM crdb_internal.kv_flow_controller AS c
           INNER JOIN crdb_internal.kv_flow_control_handles AS hs ON
             c.tenant_id = hs.tenant_id AND c.store_id = hs.tenant_id
     WHERE available_regular_tokens <= 0;

Or if looking to understand how many active replication streams each
leaseholder+leader range is shaping write traffic through, something
like:

    SELECT range_id, count(*) AS streams
      FROM crdb_internal.kv_flow_control_handles
  GROUP BY (range_id)
  ORDER BY streams DESC;

Release note: None

irfansharif committed:
admission: move CreateTime-sequencing below-raft

We move kvflowsequencer.Sequencer and its use in kvflowhandle.Handle
(above-raft) to admission.sequencer, now used by
admission.StoreWorkQueue (below-raft). This variant appeared in an
earlier revision of #97599 where we first introduced monotonically
increasing CreateTimes for a given raft group.

In a subsequent commit, when integrating kvflowcontrol into the critical
path for replication traffic, we'll observe that it's quite difficult to
create sequencing CreateTimes[^1] above raft. This is because these sequence
numbers are encoded as part of the raft proposal[^2], and at
encode-time, we don't actually know what log position the proposal is
going to end up in. It's hard to explicitly guarantee that a proposal
with log-position P1 will get encoded before another with log position
P2, where P1 < P2.

Naively sequencing CreateTimes at proposal-encode-time could result in
over-admission. This is because of how we return flow tokens -- up to
some log index[^3], and how use these sequence numbers in below-raft
WorkQueues. If P2 ends up with a lower sequence number/CreateTime, it
would get admitted first, and when returning flow tokens by log
position, in specifying up-to-P2, we'll early return P1's flow tokens
despite it not being admitted. So we'd over-admit at the sender. This
is all within a <tenant,priority> pair.

[^1]: We use CreateTimes as "sequence numbers" in replication admission
      control. We want to assign each AC-queued work below-raft a
      "sequence number" for FIFO ordering within a <tenant,priority>.
      We ensure these timestamps are roughly monotonic with respect to
      log positions of replicated work by sequencing work in log
      position order.
[^2]: In kvflowcontrolpb.RaftAdmissionMeta.
[^3]: See kvflowcontrolpb.AdmittedRaftLogEntries.

Release note: None

irfansharif committed:
admission: add intercept points for when replicated work gets admitted

In a subsequent commit, when integrating kvflowcontrol into the critical
path for replication traffic, we'll set up the return of flow tokens
from the receiver node back to the sender once log entries get
(asynchronously) admitted[^1]. So we need to intercept the exact points
at which the virtually enqueued work items get admitted, since it all
happens asynchronously[^2]. To that end we introduce the following
interface:

    // OnLogEntryAdmitted is used to observe the specific entries
    // (identified by rangeID + log position) that were admitted. Since
    // admission control for log entries is asynchronous/non-blocking,
    // this allows callers to do requisite post-admission
    // bookkeeping.
    type OnLogEntryAdmitted interface {
     AdmittedLogEntry(
       origin roachpb.NodeID, /* node where the entry originated */
       pri admissionpb.WorkPriority, /* admission priority of the entry */
       storeID roachpb.StoreID, /* store on which the entry was admitted */
       rangeID roachpb.RangeID, /* identifying range for the log entry */
       pos LogPosition, /* log position of the entry that was admitted*/
     )
    }

For now we pass in a no-op implementation in production code, but this
will change shortly.

Seeing as how the asynchronous admit interface is going to be the
primary once once we enable replication admission control by default,
for IO control, we no longer need the storeWriteDone interfaces and
corresponding types. It's being used by our current (and soon-to-be
legacy) above-raft IO admission control to inform granters of when the
write was actually done, post-admission. For above-raft IO control, at
admit-time we do not have sizing info for the writes, so by intercepting
these writes at write-done time we're able to make any outstanding token
adjustments in the granter.

To reflect this new world, we:
- Rename setAdmittedDoneModels to setLinearModels.
- Introduce a storeReplicatedWorkAdmittedInfo[^3]. It provides
  information about the size of replicated work once it's admitted
  (which happens asynchronously from the work itself). This lets us use
  the underlying linear models for L0 {writes,ingests} to deduct an
  appropriate number of tokens from the granter, for the admitted work
  size[^4].
- Rename the granterWithStoreWriteDone interface to
  granterWithStoreReplicatedWorkAdmitted. We'll still intercept the
  actual point of admission for some token adjustments, through the
  the storeReplicatedWorkAdmittedLocked API shown below. There are two
  callstacks through which this API gets invoked, one where the coord.mu
  is already held, and one where it isn't. We plumb this information
  through so the lock is acquired if not already held. The locking
  structure is unfortunate, but this was a minimally invasive diff.

   storeReplicatedWorkAdmittedLocked(
    originalTokens int64,
    admittedInfo storeReplicatedWorkAdmittedInfo,
   ) (additionalTokens int64)

While here, we also export an admission.TestingReverseWorkPriorityDict.
There are at least three tests that have re-invented the wheel.

[^1]: This will happen through the kvflowcontrol.Dispatch interface
      introduced back in #97766, after integrating it with the RaftTransport
      layer.
[^2]: Introduced in #97599, for replicated write work.
[^3]: Identical to the previous StoreWorkDoneInfo.
[^4]: There's a peculiarity here in that at enqueuing-time we actually
      know the size of the write, so we could have deducted the right
      number of tokens upfront and avoid this post-admit granter token
      adjustment. We inherit this structure from earlier, and just leave
      a TODO for now.

Release note: None

irfansharif committed:
kvserver: setup flow token returns using the RaftTransport

This commit integrates the kvflowcontrol.Dispatch with the
kvserver-level RaftTransport. When log entries are admitted below raft,
we'll want to inform the origin nodes of this fact, effectively
returning the flow tokens that were deducted when replicating the log
entry to us. We repurpose the existing RaftTransport for this
communication -- we piggyback these flow token returns[^1] on raft
messages already bound to nodes we're returning tokens to. We also
guarantee delivery of tokens in the presence of idle RaftTransport
connections[^2].

We had to to introduce some protocol changes here. When a client
establishes a RaftMessageRequestBatch stream, it sends along to the
server the set of all StoreIDs it has. It's populated on the first
RaftMessageRequestBatch sent along MultiRaft.RaftMessageBatch gRPC
stream identifying at least one store, and then populated once more if
any additional stores have been initialized[^3]. This data is used by
the kvflowcontrol machinery to track the exact set of stores on the
client node. It uses this information to react to the gRPC streams
breaking. Since these streams are used to piggy information about which
log entries were admitted below raft[^4] in order for the server-side to
free up flow tokens, if the stream breaks we possibly risk leaking these
tokens. So when these streams break, we use information about the
client's stores to release all held tokens[^5].

We're not using this code just yet, which is just the below-raft
integration with kvflowcontrol. The subsequent commit will introduce the
above-raft integration where we'll actually deduct flow tokens at the
sender, encode proposals using EntryEncoding{Standard,Sideloaded}WithAC,
which in turn enqueues virtual work items in below-raft admission queues
for asynchronous admission. Once asynchronously admitted, using the
changes in this commit, we'll return flow tokens using the now-wired-up
kvflowcontrol.Dispatch interface.

---

Suggested reading order for reviewers:

- (*RaftTransport).kvflowControl
  Brief comment block which tries to give a lay of the land.

- flow_control_stores.go
  Integration interface+implementation that's going to be used by the
  RaftTransport to return flow tokens to the specific locally held
  kvflowcontrol.Handles, after learning about admitted raft log entries
  from remote nodes. It's implemented more fully in the subsequent commit.

- flow_control_raft_transport.go
  Contains the set of new dependencies now used in the RaftTransport
  code for flow token purposes. It also includes the interfaces that
  show how the RaftTransport informs individual replicas that its no
  longer connected to specific (remote) stores. They're used more fully
  in the subsequent commit.

- raft_transport.go
  The actual code changes to the RaftTransport.

- flow_token_transport_test.go and flow_token_transport/*
  Datadriven test to understand how the various pieces fit together.

- kvflowdispatch/*
  Adds some metrics and unit testing for the canonical
  kvflowcontrol.Dispatch implementation (previously implemented).

---

[^1]: In the form of kvflowcontrolpb.AdmittedRaftLogEntries.
[^2]: See kvserver.TestFlowTokenTransport.
[^3]: This two-step process is because of how and when we allocate
      StoreIDs. Ignoring nodes that are bootstrapping the cluster (which
      just picks the initial set of StoreIDs -- see
      pkg/server.bootstrapCluster), whenever a new node is added, it's
      assigned a node ID and store ID by an existing node in CRDB (see
      kvpb.JoinNodeResponse). Subsequent store IDs, for multi-store
      nodes, are generated by the joining node by incrementing a
      sequence ID generator (see
      pkg/server.(*Node).initializeAdditionalStores). All of which is to
      say that the very first time we issue a RaftMessageRequestBatch,
      we might not have all the StoreIDs. But we will very shortly
      after, and certainly before any replicas get allocated to the
      additional store.
[^4]: See kvflowcontrolpb.AdmittedRaftLogEntries and its use in
      RaftMessageRequest.
[^5]: See I1 from kvflowcontrol/doc.go.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/101002)
<!-- Reviewable:end -->

irfansharif committed:
kvflowcontrol: add tests for Inspect() APIs

Release note: None

irfansharif committed:
kvserver: add datadriven test for kvflowcontrol integration

We add TestFlowControlIntegration, exercising the kvflowcontrol
integration interface (replicaFlowControlIntegration) introduced in an
earlier commit. It tests how the underlying kvflowcontrol.Handle is
constructed/destroyed, and its stream-management APIs invoked, as
replicas acquire raft leadership, lost it, observe paused and/or
inactive followers, change range descriptors, and react to raft progress
updates. We can write tests of the following form:

    # Observe how the integration layer deals with paused followers.
    # Start off with a triply replicated range r1/t1, with replicas on
    # n1/s1, n2/s2, and n3/s3 (with replica IDs 1-3 respectively).
    init tenant=t1 range=r1 replid=1
    ----

    state descriptor=(1,2,3) applied=1/10
    ----

    # Set up replid=1 (declared in init above) to be the raft leader. It
    # should connect to all three replication streams.
    integration op=became-leader
    ----
    initialized flow control handle for r1/t1
    connected to replication stream t1/s1 starting at log-position=1/10
    connected to replication stream t1/s2 starting at log-position=1/10
    connected to replication stream t1/s3 starting at log-position=1/10

    # Pause replid=2. Observe that we disconnect the stream to t1/s2.
    state descriptor=(1,2,3) paused=(2)
    ----

    integration op=followers-paused
    ----
    disconnected from replication stream t1/s2

These are still unit tests, testing the interface at just the
interface-level. We'll introduce more end-to-end integration testing of
an actual replica's lifecycle in subsequent commits.

Release note: None

irfansharif committed:
kvserver: add end-to-end TestFlowControl* tests

TestFlowControl* are end-to-end tests of the kvflowcontrol
machinery, replicating + admitting individual writes. They make use of
an echotest-backed test harness to make it possible to observe
sophisticated KV interactions. We can now write tests that look like
this:

    -- Flow token metrics, before issuing the regular 1MiB replicated
    -- write.
    SELECT name, crdb_internal.humanize_bytes(value::INT8)
        FROM crdb_internal.node_metrics
       WHERE name LIKE '%kvadmission%tokens%'
    ORDER BY name ASC;

      kvadmission.flow_controller.elastic_tokens_available   | 0 B
      kvadmission.flow_controller.elastic_tokens_deducted    | 0 B
      kvadmission.flow_controller.elastic_tokens_returned    | 0 B
      kvadmission.flow_controller.elastic_tokens_unaccounted | 0 B
      kvadmission.flow_controller.regular_tokens_available   | 0 B
      kvadmission.flow_controller.regular_tokens_deducted    | 0 B
      kvadmission.flow_controller.regular_tokens_returned    | 0 B
      kvadmission.flow_controller.regular_tokens_unaccounted | 0 B

    -- Flow token metrics from n1 after issuing the regular 1MiB
    -- replicated write, and it being admitted on n1, n2 and n3. We
    -- should see 3*1MiB = 3MiB of {regular,elastic} tokens deducted and
    -- returned, and {8*3=24MiB,16*3=48MiB} of {regular,elastic} tokens
    -- available. Everything should be accounted for.
    SELECT name, crdb_internal.humanize_bytes(value::INT8)
        FROM crdb_internal.node_metrics
       WHERE name LIKE '%kvadmission%tokens%'
    ORDER BY name ASC;

      kvadmission.flow_controller.elastic_tokens_available   | 24 MiB
      kvadmission.flow_controller.elastic_tokens_deducted    | 3.0 MiB
      kvadmission.flow_controller.elastic_tokens_returned    | 3.0 MiB
      kvadmission.flow_controller.elastic_tokens_unaccounted | 0 B
      kvadmission.flow_controller.regular_tokens_available   | 48 MiB
      kvadmission.flow_controller.regular_tokens_deducted    | 3.0 MiB
      kvadmission.flow_controller.regular_tokens_returned    | 3.0 MiB
      kvadmission.flow_controller.regular_tokens_unaccounted | 0 B
    ----
    ----

Release note: None

irfansharif committed:
kvserver,kvflowcontrol: integrate flow control

Part of #95563. This PR integrates various kvflowcontrol components into
the critical path for replication traffic. It does so by introducing two
"integration interfaces" in the kvserver package to intercept various
points of a replica's lifecycle, using it to manage the underlying
replication streams and flow tokens. The integration is mediated through
two cluster settings:

  - kvadmission.flow_control.enabled
      This is a top-level kill-switch to revert to pre-kvflowcontrol
      behavior where follower writes unilaterally deducted IO tokens
      without blocking.

  - kvadmission.flow_control.mode
      It can take on one of three settings, each exercising the flow
      control machinery to varying degrees.

      - apply_to_elastic
          Only applies admission delays to elastic traffic.

      - apply_to_all
          Applies admission delays to {regular,elastic} traffic.

When the mode is changed, we simply admit all waiting requests. This
risks possibly over-admitting work, but that's ok -- we assume these
mode changes are rare events and done under supervision. These
settings are hooked into in the kvadmission and kvflowcontroller
packages. As for the actual integration interfaces in kvserver, they
are:

  - replicaFlowControlIntegration: used to integrate with replication
    flow control. It's intercepts various points in a replica's
    lifecycle, like it acquiring raft leadership or losing it, or its
    raft membership changing, etc. Accessing it requires Replica.mu to
    be held, exclusively (this is asserted on in the canonical
    implementation).

      type replicaFlowControlIntegration interface {
        handle() (kvflowcontrol.Handle, bool)
        onBecameLeader(context.Context)
        onBecameFollower(context.Context)
        onDescChanged(context.Context)
        onFollowersPaused(context.Context)
        onReplicaDestroyed(context.Context)
        onProposalQuotaUpdated(context.Context)
      }

  - replicaForFlowControl abstracts the interface of an individual
    Replica, as needed by replicaFlowControlIntegration.

      type replicaForFlowControl interface {
        assertLocked()
        annotateCtx(context.Context) context.Context
        getTenantID() roachpb.TenantID
        getReplicaID() roachpb.ReplicaID
        getRangeID() roachpb.RangeID
        getDescriptor() *roachpb.RangeDescriptor
        pausedFollowers() map[roachpb.ReplicaID]struct{}
        isFollowerActive(context.Context, roachpb.ReplicaID) bool
        appliedLogPosition() kvflowcontrolpb.RaftLogPosition
        withReplicaProgress(f func(roachpb.ReplicaID, rafttracker.Progress))
      }

Release note: None

irfansharif commented:
Pulled these commits back into https://github.com/cockroachdb/cockroach/pull/98308.
